### ISSUE TITLE
NSwagStudio: Signed installer?

### ISSUE THREAD

**Issue by JeroMiya** [RicoSuter/NSwag#2906]
Would love to see a signed installer for NSwagStudio. Chrome and Windows Defender try really hard to prevent people from installing the current one.

**Comment by eluchsinger** [comment]
Any news here?

**Comment by jeremyVignelles** [comment]
Pretty sure there wasn't any progress. What is needed to do that?

**Comment by eluchsinger** [comment]
The application often gets blocked by Windows Defender because it's missing the Author's information. Let's take a look at what's needed to sign the application.

**Comment by jeremyVignelles** [comment]
If it requires to pay for a trusted code signing certificate, that's probably a no-go for many open-source projects. Plus, I assume there are some implications in the release process.

**Comment by JeroMiya** [comment]
It's not uncommon for open source projects to have signed binaries. I see two routes used:

1. [For larger projects] Some kind of foundation or non-business entity is created, and the certificate is done in their name.
2. [For smaller projects] Primary maintainer purchases and manages the certificate in their own name or an LLC in their name etc...

**Comment by jeremyVignelles** [comment]
What is the cost of such certificate? do you have any links ?

**Comment by JeroMiya** [comment]
I wasn't concerned so much as a user of the software (at least not directly), but in terms of project health. An unsigned installer is likely to dissuade users from installing it. Then fewer users generally means fewer contributors, etc...

**Comment by jeremyVignelles** [comment]
I would agree if it was a software for everyone, but as a tool for developers, I don't mind, because:
- The tool can be used without NSwagStudio
- Developers know that GitHub is a trustworthy website
- If someone has a doubt, they can clone the repo and build that themself. If they are .net devs, they have the tools for that.

No strong opposition to get it implemented though.

**Comment by eluchsinger** [comment]
@jeremyVignelles  it's not just about the developer. This tool gets flagged by different antivirus vendors - including Microsoft's Advanced Threat Protection and SmartScreen. And many developers are having issues that are not just "annoying", but actually they can't start the tool after an update for example.

I'm sure we can find a way to pay for the $70 expenses.

**Comment by RicoSuter** [comment]
Money is not the issue here - we have funds ftom donations and i’m sure Microsoft will provide one if we ask. It just was not a priority. Another option for me is to put this in the MS App Store, would this also be ok?

**Comment by jeremyVignelles** [comment]
As long as Microsoft store is not mandatory, I'm fine with that.
I like to be able to install multiple projects side by side by unzipping the folder, and being able to choose the version i'm on, as updates are frequent in this projects, but updates in my code takes time.

**Comment by RicoSuter** [comment]
> As long as Microsoft store is not mandatory, I'm fine with that.

Sure, the regular installer will still be available - store is only needed for auto updates and signed install (maybe with a small price :-)

> and being able to choose the version i'm on, as updates are frequent in this projects, but updates in my code takes time.

That's why i recommend to use NSwag via NPM or NuGet (NSwag.MSBuild) as this way you can control the version per project.

**Comment by eluchsinger** [comment]
@RicoSuter again, I am facing the same issues. The built-in Update mechanism can not download the latest version (it times out, probably blocked by Windows Defender). Then after downloading the .MSI, I get warnings on download, after download and before executing. After installation, NSwagStudio does not start because it's process is blocked by Windows Defender, making it look like a bug in the application because the user does not get any notification at all. It just doesn't start.

Now I know about these issues. I have had to help many colleagues multiple times, because they don't know about these issues at first and it's basically just luck finding out about why the application does not start.

It would be nice to have this issue fixed and for me it's actually one with a big impact. If you want to publish to Microsoft Store, even better, as it would make it even easier to install and update.

**Comment by RicoSuter** [comment]
I agree we should sign the app either on the MSI or by publishing to the store...

However I only use the UI (NSwagStudio) for initial setup of the config file (nswag.json) and from then on everything is either automated (auto gen) or executed via CLI... so other people working on the project never need to start the UI to manually generate clients or spec etc.

**Comment by eluchsinger** [comment]
@RicoSuter I agree with the CLI. I have implemented that approach on a different project and it seems to work fine. I will migrate this project to do just that aswell. Still it would be nice to sign everything as NSwag is very popular and users should be able to trust it.

**Comment by psandgren** [comment]
This is still a problem that would be awesome to solve by signing the files. Makes it hard to use in an enterprise org with stasi rules.

But, as a quickfix, maybe add a little more to the [guide](https://github.com/RicoSuter/NSwag/wiki/NSwag.MSBuild) how to add clients to project when the initial config is done.
